### PR TITLE
Webhooks for AntreaConfig and CalicoConfig

### LINKED ISSUE #1690: Webhook for AntreaConfig and CalicoConfig
- AntreaConfig immutable fields are TLSCipherSuites, DefaultMTU, TrafficEncapMode, NoSNAT, DisableUDPTunnelOffload
- CalicoConfig immutable fields are namespace and VethMTU

### PR DESCRIPTION
**Pull request by shivaani0505**
### What this PR does / why we need it
- Adding webhooks for AntreaConfig and CalicoConfig
- Made the webhook and certificate tests self-contained to avoid unwanted side-effects
- Added webhook manifests for CNI (Antrea and Calico)
- Added tests for validating and mutating webhooks for Antrea
- Added tests for webhooks for Calico

### Which issue(s) this PR fixes

Fixes #1690 

### Release note

```release-note
package-based-lcm: Add validating webhooks for AntreaConfig and CalicoConfig
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

#### Special notes for your reviewer

I have to add manifests file as well, which I generated locally. As of now just using the manifests in testdata for Validating and Mutating webhooks

### CODE REVIEW

File: addons/controllers/antreaconfig_controller_test.go
Comment (danniel1205) [inline addons/controllers/antreaconfig_controller_test.go:22]: blocking: we might want to change back to `Describe` before merge.

File: addons/controllers/calicoconfig_controller_test.go
Comment (danniel1205) [inline addons/controllers/calicoconfig_controller_test.go:24]: blocking: we might want to change back to `Describe` before merge.

File: addons/controllers/antreaconfig_controller_test.go
Comment (shivaani0505) [reply addons/controllers/antreaconfig_controller_test.go:22]: I don't plan to check this in. This is to only facilitate testing for time being

File: addons/controllers/calicoconfig_controller_test.go
Comment (shivaani0505) [reply addons/controllers/calicoconfig_controller_test.go:24]: I don't plan to check this in. This is to only facilitate testing for time being

File: addons/controllers/suite_test.go
Comment (danniel1205) [inline addons/controllers/suite_test.go:286]: suggestion: is it possible we put the following webhook setup into a separate function ? `BeforeSuite` holds 200+ lines which makes the maintenance hard

File: addons/controllers/suite_test.go
Comment (danniel1205) [inline addons/controllers/suite_test.go:40]: question: does linter complains if removing above newline ?

File: apis/cni/v1alpha1/antreaconfig_webhook.go
Comment (danniel1205) [inline apis/cni/v1alpha1/antreaconfig_webhook.go:28]: suggestion: remove above before merging

**Review by danniel1205 — CHANGES_REQUESTED** [review]
The `zz_deepcopy` file has be modified, we would like to double check if that change is expected before merge. 

And `BeforeSuite` holds 200+ lines in a single function, it would be great if we could split things up in to separate functions.

File: addons/controllers/suite_test.go
Comment (blc1996) [reply addons/controllers/suite_test.go:286]: Agreed, we should make the webhook setup related code (starting from line 280) as a separate function

File: addons/controllers/suite_test.go
Comment (blc1996) [inline addons/controllers/suite_test.go:290]: Will this line of code affect the test case in `certificates_test.go`?

File: apis/cni/v1alpha1/antreaconfig_webhook.go
Comment (vijaykatam) [reply apis/cni/v1alpha1/antreaconfig_webhook.go:36]: If default is doing nothing lets remove this and `var _ webhook.Defaulter = &AntreaConfig{}`

File: apis/cni/v1alpha1/antreaconfig_webhook.go
Comment (vijaykatam) [inline apis/cni/v1alpha1/antreaconfig_webhook.go:88]: @lubronzhan @jayunit100 - Do you know if antrea defaultMTU can be changed after cluster create?

File: apis/cni/v1alpha1/antreaconfig_webhook.go
Comment (vijaykatam) [inline apis/cni/v1alpha1/antreaconfig_webhook.go:131]: remove log for no op

File: apis/cni/v1alpha1/calicoconfig_webhook.go
Comment (vijaykatam) [inline apis/cni/v1alpha1/calicoconfig_webhook.go:34]: same comment as above, remove default if this is no-op

File: apis/cni/v1alpha1/calicoconfig_webhook.go
Comment (vijaykatam) [inline apis/cni/v1alpha1/calicoconfig_webhook.go:44]: remove log for no-op

File: apis/cni/v1alpha1/calicoconfig_webhook.go
Comment (vijaykatam) [inline apis/cni/v1alpha1/calicoconfig_webhook.go:67]: @lubronzhan @jayunit100 - Same question as antrea. Can vethMTU for calico be changed after cluster creation?

File: apis/cni/v1alpha1/calicoconfig_webhook.go
Comment (vijaykatam) [inline apis/cni/v1alpha1/calicoconfig_webhook.go:83]: remove log for no-op

File: apis/cni/v1alpha1/antreaconfig_webhook.go
Comment (vijaykatam) [reply apis/cni/v1alpha1/antreaconfig_webhook.go:28]: blocker: cleanup

File: addons/controllers/suite_test.go
Comment (vijaykatam) [reply addons/controllers/suite_test.go:286]: +1 a separate helper function for webhook setup

**Review by vijaykatam — COMMENTED** [review]
Remove release note as this should not show up in t-f release notes yet. Additionally integration with main.go behind feature flag needs to happen.

File: addons/controllers/testdata/test-webhook-manifests.yaml
Comment (blc1996) [inline addons/controllers/testdata/test-webhook-manifests.yaml:9]: Trying to understand the `externalName: 127.0.0.1` here. Is it trying to redirect traffic to the webhook server started by mgr.GetWebhookServer()?
If so, I this we should probably explicitly configure mgr.GetWebhookServer() in suite_setup.

File: apis/cni/v1alpha1/zz_generated.deepcopy.go
Comment (shivaani0505) [reply apis/cni/v1alpha1/zz_generated.deepcopy.go:12]: I think this change is harmless. This change was generated as a part of procedure for generating manifests.
You can use import  "runtime "k8s.io/apimachinery/pkg/runtime" or like "k8s.io/apimachinery/pkg/runtime"

File: addons/controllers/suite_test.go
Comment (shivaani0505) [reply addons/controllers/suite_test.go]: Fixed newlines. No linter complain

File: apis/cni/v1alpha1/antreaconfig_webhook.go
Comment (shivaani0505) [reply apis/cni/v1alpha1/antreaconfig_webhook.go]: Fixed

File: apis/cni/v1alpha1/calicoconfig_webhook.go
Comment (shivaani0505) [reply apis/cni/v1alpha1/calicoconfig_webhook.go:34]: Fixed

File: apis/cni/v1alpha1/antreaconfig_webhook.go
Comment (shivaani0505) [reply apis/cni/v1alpha1/antreaconfig_webhook.go]: Fixed

File: apis/cni/v1alpha1/calicoconfig_webhook.go
Comment (shivaani0505) [reply apis/cni/v1alpha1/calicoconfig_webhook.go]: Fixed

File: apis/cni/v1alpha1/calicoconfig_webhook.go
Comment (shivaani0505) [reply apis/cni/v1alpha1/calicoconfig_webhook.go]: Fixed

File: addons/controllers/testdata/test-webhook-manifests.yaml
Comment (shivaani0505) [reply addons/controllers/testdata/test-webhook-manifests.yaml:9]: Though it is the default setting, but we are still explicitly setting these values to avoid any side effects.

File: addons/controllers/suite_test.go
Comment (shivaani0505) [reply addons/controllers/suite_test.go]: I made each test self-contained, so this should not affect the certificate_test.go. The test cases in certificate_test.go independently setup the certificates when run.

File: addons/controllers/suite_test.go
Comment (shivaani0505) [reply addons/controllers/suite_test.go]: Fixed. Added a helper function and refactored all the affected tests

**Comment by shivaani0505** [comment]
> The `zz_deepcopy` file has be modified, we would like to double check if that change is expected before merge.
> 
> And `BeforeSuite` holds 200+ lines in a single function, it would be great if we could split things up in to separate functions.

Fixed by introducing helper function

File: addons/controllers/testdata/test-antrea-calico-webhook-manifests.yaml
Comment (blc1996) [inline addons/controllers/testdata/test-antrea-calico-webhook-manifests.yaml:5]: Generic service name might be mixed up with other webhooks. Better to give it a name related to Antrea and Calico.

File: addons/controllers/testdata/test-antrea-calico-webhook-manifests.yaml
Comment (blc1996) [inline addons/controllers/testdata/test-antrea-calico-webhook-manifests.yaml:17]: Should we add CNI related info in this naming?

File: addons/controllers/testdata/test-antrea-calico-webhook-manifests.yaml
Comment (blc1996) [inline addons/controllers/testdata/test-antrea-calico-webhook-manifests.yaml:31]: Probably update the name here?

File: addons/controllers/calicoconfig_controller_test.go
Comment (blc1996) [inline addons/controllers/calicoconfig_controller_test.go:33]: Code under `JustBeforeEach ` will be executed before each test case. It will complain resource already exist if there is more than one testcase. Consider move it somewhere else, in case we add more test cases in the future.

File: addons/controllers/testdata/test-antrea-calico-webhook-manifests.yaml
Comment (shivaani0505) [reply addons/controllers/testdata/test-antrea-calico-webhook-manifests.yaml:5]: As of now, the design is for a common webhook server and webhook service that allso share the same certificates. Hence the generic name

File: addons/controllers/calicoconfig_controller_test.go
Comment (shivaani0505) [reply addons/controllers/calicoconfig_controller_test.go:33]: Yes the resources are created in `JustBeforeEach` but they are also deleted in `AfterEach`

File: apis/cni/v1alpha1/antreaconfig_webhook.go
Comment (jayunit100) [reply apis/cni/v1alpha1/antreaconfig_webhook.go:85]: @yzaccc can you make the final call on this one ? thanks

Comment (jayunit100) [reply apis/cni/v1alpha1/antreaconfig_webhook.go:85]: Zac is on PTO, but MTU can be modified in antrea.  In the case that you set it, and delete an `antrea-agent`, the new agent will have a new MTU. 

Im thus going to say it should allow mutation.  In general i think all params should allow mutation unless we specify otherwise - -

File: apis/cni/v1alpha1/calicoconfig_webhook.go
Comment (jayunit100) [reply apis/cni/v1alpha1/calicoconfig_webhook.go:63]: Yes in general the VethMTU for calico can change.  in this case, calico will automatically pick the new MTU up for you https://projectcalico.docs.tigera.io/networking/mtu

File: apis/cni/v1alpha1/antreaconfig_webhook.go
Comment (shivaani0505) [reply apis/cni/v1alpha1/antreaconfig_webhook.go:85]: Thanks for the info. will remove the mutation check as it is allowed

File: apis/cni/v1alpha1/calicoconfig_webhook.go
Comment (shivaani0505) [reply apis/cni/v1alpha1/calicoconfig_webhook.go:63]: Thanks for the info. will remove the mutation check as it is allowed

File: apis/cni/v1alpha1/antreaconfig_webhook.go
Comment (shivaani0505) [reply apis/cni/v1alpha1/antreaconfig_webhook.go:85]: @jayunit100 , so are there any immutable fields for antrea and calico?

**Comment by shivaani0505** [comment]
Webhook Integration with main will be done by follow up PR (https://github.com/vmware-tanzu/tanzu-framework/issues/1851)

**Review by vijaykatam — APPROVED** [review]

**Review by danniel1205 — APPROVED** [review]

File: apis/cni/v1alpha1/antreaconfig_webhook.go
Comment (yzaccc) [reply apis/cni/v1alpha1/antreaconfig_webhook.go]: Looks like this is already merged. But yea, MTU is mutable, should not be checked here.

**Pull request merged**
